File: clients/android/app/src/main/kotlin/com/lextures/android/features/quiz/CodeQuestionView.kt
package com.lextures.android.features.quiz

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.material3.CircularProgressIndicator
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.rememberCoroutineScope
import androidx.compose.runtime.setValue
import androidx.compose.ui.Modifier
import androidx.compose.ui.text.font.FontFamily
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import com.lextures.android.core.design.AuthPrimaryButton
import com.lextures.android.core.design.LexturesColors
import com.lextures.android.core.i18n.L
import com.lextures.android.core.lms.LmsApi
import com.lextures.android.core.lms.QuizAnswerState
import com.lextures.android.core.lms.QuizCodeRunResponse
import com.lextures.android.core.lms.QuizLogic
import com.lextures.android.core.lms.QuizQuestion
import com.lextures.android.core.ui.CodeEditor
import com.lextures.android.R
import kotlinx.coroutines.launch

data class CodeQuestionRunContext(
    val courseCode: String,
    val itemId: String,
    val attemptId: String,
    val accessToken: String,
)

@Composable
fun CodeQuestionView(
    question: QuizQuestion,
    answer: QuizAnswerState,
    runContext: CodeQuestionRunContext?,
    onChange: (QuizAnswerState) -> Unit,
    modifier: Modifier = Modifier,
) {
    if (QuizLogic.isCodeQuestionOversized(question)) {
        Column(modifier = modifier.padding(vertical = 4.dp)) {
            Text(quizCodeOversizedTitle(), fontWeight = FontWeight.SemiBold)
            Text(quizCodeOversizedHint(), color = LexturesColors.TextSecondary)
        }
        return
    }

    val scope = rememberCoroutineScope()
    val runFailedMessage = quizCodeRunFailedLabel()
    var running by remember(question.id) { mutableStateOf(false) }
    var runError by remember(question.id) { mutableStateOf<String?>(null) }
    var runResult by remember(question.id) { mutableStateOf<QuizCodeRunResponse?>(null) }
    var seeded by remember(question.id) { mutableStateOf(false) }
    val language = QuizLogic.codeLanguageLabel(question)

    LaunchedEffect(question.id) {
        if (!seeded) {
            seeded = true
            val initial = QuizLogic.initialCodeAnswer(question, answer)
            if (initial != answer) onChange(initial)
        }
    }

    Column(modifier = modifier, verticalArrangement = Arrangement.spacedBy(10.dp)) {
        Text(L.format(R.string.mobile_quiz_code_language, language), fontWeight = FontWeight.Medium)

        CodeEditor(
            text = answer.text.orEmpty(),
            onTextChange = { onChange(answer.copy(text = it)) },
            onInsert = { snippet -> onChange(answer.copy(text = answer.text.orEmpty() + snippet)) },
        )

        AuthPrimaryButton(
            text = if (running) quizCodeRunningLabel() else quizCodeRunLabel(),
            onClick = {
                val ctx = runContext ?: return@AuthPrimaryButton
                scope.launch {
                    running = true
                    runError = null
                    try {
                        runResult = LmsApi.postQuizQuestionRun(
                            courseCode = ctx.courseCode,
                            itemId = ctx.itemId,
                            attemptId = ctx.attemptId,
                            questionId = question.id,
                            code = answer.text.orEmpty(),
                            languageId = question.typeConfig?.languageId,
                            accessToken = ctx.accessToken,
                        )
                    } catch (e: Exception) {
                        runError = e.message ?: runFailedMessage
                    } finally {
                        running = false
                    }
                }
            },
            enabled = !running && runContext != null && !answer.text.isNullOrBlank(),
            modifier = Modifier.fillMaxWidth(),
        )

        if (running) {
            CircularProgressIndicator()
        }

        runError?.let {
            Text(it, color = LexturesColors.Coral)
        }

        runResult?.let { response ->
            CodeRunResultsPanel(response)
        }
    }
}

@Composable
private fun CodeRunResultsPanel(response: QuizCodeRunResponse) {
    Column(verticalArrangement = Arrangement.spacedBy(8.dp)) {
        Text(L.format(R.string.mobile_quiz_code_runScore, response.pointsEarned, response.pointsPossible))
        response.results.forEachIndexed { index, result ->
            Column(
                Modifier
                    .fillMaxWidth()
                    .padding(10.dp),
                verticalArrangement = Arrangement.spacedBy(4.dp),
            ) {
                Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
                    Text(L.format(R.string.mobile_quiz_code_testNumber, index + 1), fontWeight = FontWeight.SemiBold)
                    Text(
                        quizCodeStatusLabel(result.status),
                        color = if (result.passed) LexturesColors.Primary else LexturesColors.Coral,
                        fontWeight = FontWeight.Bold,
                    )
                }
                if (result.expectedOutput.isNotEmpty()) {
                    Text(
                        L.format(R.string.mobile_quiz_code_expected, result.expectedOutput),
                        fontFamily = FontFamily.Monospace,
                    )
                }
                Text(
                    L.format(R.string.mobile_quiz_code_actual, result.actualOutput),
                    fontFamily = FontFamily.Monospace,
                )
                result.stderr?.takeIf { it.isNotEmpty() }?.let { stderr ->
                    Text(
                        L.format(R.string.mobile_quiz_code_stderr, stderr),
                        fontFamily = FontFamily.Monospace,
                        color = LexturesColors.Coral,
                    )
                }
            }
        }
    }
}

@Composable
private fun quizCodeStatusLabel(status: String): String = when (status) {
    "pass" -> quizCodeStatusPassLabel()
    "fail" -> quizCodeStatusFailLabel()
    "tle" -> quizCodeStatusTleLabel()
    "mle" -> quizCodeStatusMleLabel()
    "re" -> quizCodeStatusReLabel()
    "ce" -> quizCodeStatusCeLabel()
    else -> status.uppercase()
}

File: clients/ios/Lextures/Features/Quiz/CodeQuestionView.swift
import SwiftUI

struct CodeQuestionRunContext: Equatable {
    let courseCode: String
    let itemId: String
    let attemptId: String
    let accessToken: String
}

struct CodeQuestionView: View {
    @Environment(\.colorScheme) private var colorScheme

    let question: QuizQuestion
    let answer: QuizAnswerState
    let runContext: CodeQuestionRunContext?
    let onChange: (QuizAnswerState) -> Void

    @State private var running = false
    @State private var runError: String?
    @State private var runResult: QuizCodeRunResponse?
    @State private var seededStarter = false

    private var language: String { QuizLogic.codeLanguageLabel(for: question) }

    var body: some View {
        if QuizLogic.isCodeQuestionOversized(question) {
            oversizedCard
        } else {
            editorBody
        }
    }

    private var oversizedCard: some View {
        VStack(alignment: .leading, spacing: 8) {
            Label(L.text("mobile.quiz.code.oversizedTitle"), systemImage: "macwindow")
                .font(.subheadline.weight(.semibold))
                .foregroundStyle(LexturesTheme.textPrimary(for: colorScheme))
            Text(L.text("mobile.quiz.code.oversizedHint"))
                .font(.caption)
                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))
        }
        .padding(12)
        .frame(maxWidth: .infinity, alignment: .leading)
        .background(LexturesTheme.amber.opacity(0.1))
        .clipShape(RoundedRectangle(cornerRadius: 12, style: .continuous))
    }

    private var editorBody: some View {
        VStack(alignment: .leading, spacing: 12) {
            Text(L.format("mobile.quiz.code.language", language))
                .font(.caption.weight(.semibold))
                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))

            CodeEditor(
                text: Binding(
                    get: { answer.text ?? "" },
                    set: { value in
                        var next = answer
                        next.text = value
                        onChange(next)
                    }
                ),
                language: language,
                onInsert: insertSnippet
            )

            Button {
                Task { await runTests() }
            } label: {
                HStack {
                    if running {
                        ProgressView()
                            .controlSize(.small)
                    }
                    Text(running ? L.text("mobile.quiz.code.running") : L.text("mobile.quiz.code.run"))
                        .font(.subheadline.weight(.semibold))
                }
                .frame(maxWidth: .infinity)
                .padding(.vertical, 10)
            }
            .buttonStyle(.borderedProminent)
            .tint(LexturesTheme.primary)
            .disabled(running || runContext == nil || (answer.text ?? "").trimmingCharacters(in: .whitespacesAndNewlines).isEmpty)

            if let runError {
                Text(runError)
                    .font(.caption)
                    .foregroundStyle(LexturesTheme.coral)
            }

            if let runResult {
                resultsPanel(runResult)
            }
        }
        .onAppear {
            guard !seededStarter else { return }
            seededStarter = true
            let initial = QuizLogic.initialCodeAnswer(for: question, existing: answer)
            if initial != answer {
                onChange(initial)
            }
        }
    }

    private func insertSnippet(_ snippet: String) {
        var next = answer
        let current = next.text ?? ""
        next.text = current + snippet
        onChange(next)
    }

    @MainActor
    private func runTests() async {
        guard let runContext else { return }
        running = true
        runError = nil
        defer { running = false }
        do {
            let result = try await LMSAPI.postQuizQuestionRun(
                courseCode: runContext.courseCode,
                itemId: runContext.itemId,
                attemptId: runContext.attemptId,
                questionId: question.id,
                code: answer.text ?? "",
                languageId: question.typeConfig?.languageId,
                accessToken: runContext.accessToken
            )
            runResult = result
        } catch {
            runError = (error as? LocalizedError)?.errorDescription ?? L.text("mobile.quiz.code.runFailed")
        }
    }

    @ViewBuilder
    private func resultsPanel(_ response: QuizCodeRunResponse) -> some View {
        VStack(alignment: .leading, spacing: 8) {
            Text(
                L.format(
                    "mobile.quiz.code.runScore",
                    response.pointsEarned.formatted(),
                    response.pointsPossible.formatted()
                )
            )
                .font(.caption.weight(.semibold))
                .foregroundStyle(LexturesTheme.textSecondary(for: colorScheme))

            ForEach(Array(response.results.enumerated()), id: \.offset) { index, result in
                VStack(alignment: .leading, spacing: 4) {
                    HStack {
                        Text(L.format("mobile.quiz.code.testNumber", index + 1))
                            .font(.caption.weight(.semibold))
                        Spacer()
                        Text(codeRunStatusLabel(result.status))
                            .font(.caption.weight(.bold))
                            .foregroundStyle(result.passed ? LexturesTheme.primary : LexturesTheme.coral)
                    }
                    if !result.expectedOutput.isEmpty {
                        Text(L.format("mobile.quiz.code.expected", result.expectedOutput))
                            .font(.caption2.monospaced())
                    }
                    Text(L.format("mobile.quiz.code.actual", result.actualOutput))
                        .font(.caption2.monospaced())
                    if let stderr = result.stderr, !stderr.isEmpty {
                        Text(L.format("mobile.quiz.code.stderr", stderr))
                            .font(.caption2.monospaced())
                            .foregroundStyle(LexturesTheme.coral)
                    }
                }
                .padding(10)
                .frame(maxWidth: .infinity, alignment: .leading)
                .background(LexturesTheme.sceneBackground(for: colorScheme))
                .clipShape(RoundedRectangle(cornerRadius: 10, style: .continuous))
            }
        }
        .accessibilityElement(children: .contain)
        .accessibilityLabel(L.text("mobile.quiz.code.resultsA11y"))
    }

    private func codeRunStatusLabel(_ status: String) -> String {
        switch status {
        case "pass": return L.text("mobile.quiz.code.statusPass")
        case "fail": return L.text("mobile.quiz.code.statusFail")
        case "tle": return L.text("mobile.quiz.code.statusTle")
        case "mle": return L.text("mobile.quiz.code.statusMle")
        case "re": return L.text("mobile.quiz.code.statusRe")
        case "ce": return L.text("mobile.quiz.code.statusCe")
        default: return status.uppercased()
        }
    }
}
